### PR TITLE
Add support for weak etags.

### PR DESCRIPTION
**Pull request by NavyCoat**
BTW. My first github pull request :)
relates to: #383

### CODE REVIEW

**Comment by NavyCoat** [comment]
Looks like build failed on something not related to my code changes but build config,

File: src/EventListener/CacheControlListener.php [hunk 1]
```diff
@@ -143,7 +143,11 @@ public function onKernelResponse(FilterResponseEvent $event)
         if (!empty($options['etag'])
             && ($options['overwrite'] || null === $response->getEtag())
         ) {
-            $response->setEtag(md5($response->getContent()));
+            if ($options['etag'] === 'weak') {
+                $response->setEtag(md5($response->getContent()), true);
```
Comment (dbu) [inline src/EventListener/CacheControlListener.php]: lets not do an if-else but simplify to `$response->setEtag(md5(...), 'weak' === $options['etag']);`

**Comment by NavyCoat** [comment]
@dbu It is currently 2.2 in branch-alias.

File: Resources/doc/reference/configuration/headers.rst [hunk 2]
```diff
@@ -215,7 +215,7 @@ directives are flags that are included when set to true:
 ``etag``
 """"""""
 
-**type**: ``boolean``
+**type**: ``enum`` **options**: ``true``, ``false``, ``strong``, ``weak``
```
Comment (dbu) [inline Resources/doc/reference/configuration/headers.rst]: we need to remove `true` here as its not an option anymore (it still works, but we want people to use explicit options.

lets use quotes `"` around the string options to indicate that false is boolean and not the string boolean.

File: Resources/doc/reference/configuration/headers.rst [hunk 3]
```diff
@@ -229,6 +229,22 @@ This enables a simplistic ETag calculated as md5 hash of the response body:
                     headers:
                         etag: true
 
+
+.. versionadded:: 2.2
+
+You can set up ETag to be strong or weak by passing values "strong", "weak".
```
Comment (dbu) [inline Resources/doc/reference/configuration/headers.rst]: this section needs to be indented by 4 spaces for rst to pick it up.

lets say: ... by setting the option to "strong" or "weak" respectively.

File: Resources/doc/reference/configuration/headers.rst [hunk 4]
```diff
@@ -229,6 +229,22 @@ This enables a simplistic ETag calculated as md5 hash of the response body:
...
+Value true is still treated as 'strong' due to backward compatibility.
```
Comment (dbu) [inline Resources/doc/reference/configuration/headers.rst]: lets remove this line here, this should only be in the code and the CHANGELOG, but not in the doc.

File: Resources/doc/reference/configuration/headers.rst [hunk 5]
```diff
@@ -245,6 +261,9 @@ This enables a simplistic ETag calculated as md5 hash of the response body:
     page. If the page depends on permissions, make sure to make the ETag differ
     based on those permissions (e.g. by appending the :doc:`user context hash </features/user-context>`).
 
+    You may want to set your ETags weak when you are using gzip compression on server.
```
Comment (dbu) [inline Resources/doc/reference/configuration/headers.rst]: is the server the webserver, or the reverse proxy? i would prefer to be more explicit here, maybe also say why that is, cite the RFC that talks about this or some other further info.

and "may" or "should"?

File: src/DependencyInjection/Configuration.php [hunk 6]
```diff
@@ -214,9 +214,13 @@ private function addCacheControlSection(ArrayNodeDefinition $rootNode)
...
                     ->end()
-                    ->booleanNode('etag')
+                    ->enumNode('etag')
                         ->defaultValue(false)
-                        ->info('Set a simple ETag which is just the md5 hash of the response body')
+                        ->treatTrueLike('strong')
+                        ->info('Set a simple ETag which is just the md5 hash of the response body. '.
+                               'You can specify which type of ETag you want by passing "strong" or "weak", '.
+                               'value true is treated as "strong".')
+                        ->values(['weak', 'strong', false, true])
```
Comment (dbu) [inline src/DependencyInjection/Configuration.php]: please remove true here, the treatTrueLike should take care to always translate this to "strong".

File: src/DependencyInjection/Configuration.php [hunk 7]
```diff
@@ -214,9 +214,13 @@ private function addCacheControlSection(ArrayNodeDefinition $rootNode)
...
                         ->end()
                     ->end()
-                    ->booleanNode('etag')
+                    ->enumNode('etag')
                         ->defaultValue(false)
-                        ->info('Set a simple ETag which is just the md5 hash of the response body')
+                        ->treatTrueLike('strong')
+                        ->info('Set a simple ETag which is just the md5 hash of the response body. '.
+                               'You can specify which type of ETag you want by passing "strong" or "weak", '.
+                               'value true is treated as "strong".')
```
Comment (dbu) [inline src/DependencyInjection/Configuration.php]: lets not mention this, in the changelog is good enough

**Comment by dbu** [comment]
> It is currently 2.2 in branch-alias.

sorry, i must have been confused. the branch-alias fine then!

**Comment by NavyCoat** [comment]
@dbu Okay I believe this will be last commit :P. 
BTW. When are you gonna release 2.2.0?

File: Resources/doc/reference/configuration/headers.rst [hunk 8]
```diff
@@ -245,6 +249,12 @@ This enables a simplistic ETag calculated as md5 hash of the response body:
     page. If the page depends on permissions, make sure to make the ETag differ
     based on those permissions (e.g. by appending the :doc:`user context hash </features/user-context>`).
 
+    You should want to set your ETags weak when you are using gzip compression on webserver.
```
Comment (dbu) [inline Resources/doc/reference/configuration/headers.rst]: the spelling does not like webserver. i think we opted for the more grammatically correct "web server" elsewhere in the doc. can you please adjust that?

Comment (dbu) [reply Resources/doc/reference/configuration/headers.rst]: and etags is also not found in spelling. can we say "You want to set weak ETag headers when ..."

**Comment by NavyCoat** [comment]
@dbu Okay, spelling errors fixed 👌

**Comment by dbu** [comment]
ups, fell from my radar. thanks a lot for this contribution!

**Pull request merged**